DuckDuckGo/Onboarding/ContextualOnboarding/ContextualOnboardingStateMachine.swift

```diff
@@ -0,0 +1,260 @@
+//
+//  ContextualOnboardingStateMachine.swift
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import PrivacyDashboard
+
+protocol ContextualOnboardingDialogTypeProviding {
+    func dialogTypeForTab(_ tab: Tab, privacyInfo: PrivacyInfo?) -> ContextualDialogType?
+}
+
+protocol ContextualOnboardingStateUpdater {
+    func updateStateFor(tab: Tab)
+    func gotItPressed()
+    func fireButtonUsed()
+}
+
+enum ContextualDialogType: Equatable {
+    case tryASearch
+    case searchDone(shouldFollowUp: Bool)
+    case tryASite
+    case trackers(message: NSAttributedString, shouldFollowUp: Bool)
+    case tryFireButton
+    case highFive
+}
+
+enum ContextualOnboardingState: String {
+    case notStarted
+    case showTryASearch
+    case showSearchDone
+    case showBlockedTrackers
+    case showMajorOrNoTracker
+    case showTryASite
+    case searchDoneShowBlockedTrackers
+    case searchDoneShowMajorOrNoTracker
+    case fireUsedTryASearchShown
+    case fireUsedShowSearchDone
+    case showFireButton
+    case showHighFive
+    case onboardingCompleted
+}
+
+final class ContextualOnboardingStateMachine: ContextualOnboardingDialogTypeProviding, ContextualOnboardingStateUpdater {
+
+    let trackerMessageProvider: TrackerMessageProviding
+    let startUpPreferences: StartupPreferences
+
+    @UserDefaultsWrapper(key: .contextualOnboardingState, defaultValue: ContextualOnboardingState.onboardingCompleted.rawValue)
+    private var stateString: String {
+        didSet {
+            if stateString == ContextualOnboardingState.notStarted.rawValue {
+                startUpPreferences.launchToCustomHomePage = true
+                lastVisitTab = nil
+                lastVisitSite = nil
+            }
+            if stateString == ContextualOnboardingState.onboardingCompleted.rawValue {
+                startUpPreferences.launchToCustomHomePage = false
```

Comment:
should we nil `lastVisitTab = nil` & `lastVisitSite = nil` once we complete the onboarding?

Reply:
I don’t think it should make a difference but I’ll add it